### PR TITLE
Ic pomodoro team service unit test

### PR DESCRIPTION
**Pull request by ivanCosic**

### CODE REVIEW

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+        assertEquals("TeamName in DTO should match with expected value", "TestTeam", teamDto.getName());
+        assertEquals("TeamId in DTO should match with expected value", 99, teamDto.getId());
+
+        verify(teamDaoMock, times(1)).findById(99);
+        verifyNoMoreInteractions(teamDaoMock);
+        verifyZeroInteractions(userDaoMock, userTeamDaoMock);
+    }
+
+    @Test
+    public void testFind_WrongId() {
```
Comment (stanislavgrujic) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: Pleace check the parameter`expected` for the `Test`annotation

Comment (stanislavgrujic) [reply service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: Please apply the logic to all tests where you're expecting an exception.

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        userTeam = new UserTeam();
+        userTeam.setUserTeamId(1L);
+        userTeam.setUser(user);
+        userTeam.setTeam(team);
+
+        user.getUserTeams().add(userTeam);
+    }
+
+    @Test
+    public void testMockCreation() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: Missing Javadoc

Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: test name should be something like shouldCreateMock..

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        assertNotNull("TeamDaoMock has not initialized properly", teamDaoMock);
+        assertNotNull("UserDaoMock has not initialized properly", userDaoMock);
+        assertNotNull("UserTeamDaoMock has not initialized properly", userTeamDaoMock);
+    }
+
+    @Test
+    public void testFind_CorrectId() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        assertEquals("TeamName in DTO should match with expected value", "TestTeam", teamDto.getName());
+        assertEquals("TeamId in DTO should match with expected value", 99, teamDto.getId());
+
+        verify(teamDaoMock, times(1)).findById(99);
+        verifyNoMoreInteractions(teamDaoMock);
+        verifyZeroInteractions(userDaoMock, userTeamDaoMock);
+    }
+
+    @Test(expected = EntityNotFoundException.class)
+    public void testFind_WrongId() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        given(teamDaoMock.findById(anyLong())).willThrow(new EntityNotFoundException("Team", 150L));
+        service.find(150L);
+    }
+
+    @Test
+    public void testCreateTeam_UniqueName() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        assertEquals("Returned team name should match with expected", "TestTeam", checkDTO.getName());
+
+        verify(teamDaoMock, times(1)).checkIfTeamExists(anyString());
+        verify(teamDaoMock, times(1)).save(any());
+        verifyNoMoreInteractions(teamDaoMock);
+        verifyZeroInteractions(userDaoMock, userTeamDaoMock);
+    }
+
+    @Test(expected = TeamNameNotUniqueException.class)
+    public void testCreateTeam_NotUniqueName() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        given(teamDaoMock.checkIfTeamExists("TestTeam")).willReturn(true);
+        service.createTeam("TestTeam");
+    }
+
+    @Test
+    public void testLeaveTeam_Successful() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+        verify(userDaoMock, times(1)).findById(anyLong());
+        verify(teamDaoMock, times(1)).findById(anyLong());
+        verify(userTeamDaoMock, times(1)).getByUserAndTeam(any(), any());
+        verify(userDaoMock, times(1)).update(any());
+        verifyNoMoreInteractions(userDaoMock, teamDaoMock, userTeamDaoMock);
+    }
+
+    @Test(expected = FailedToLeaveTeamException.class)
+    public void testLeaveTeam_Unsuccessful() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        given(teamDaoMock.findById(99)).willReturn(team);
+        given(userTeamDaoMock.getByUserAndTeam(user, team)).willReturn(userTeam);
+        given(userDaoMock.update(user)).willReturn(user1); //Simulate DB fail to remove userTeam from user
+
+        //when leaveTeam() throw FailedToLeaveTeamException
+        service.leaveTeam(99L, 1L);
+    }
+
+    @Test
+    public void testJoinTeam_Successful() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+        verify(userTeamDaoMock, times(1)).checkIfUserTeamExists(anyLong(), anyLong());
+        verify(teamDaoMock, times(1)).findById(anyLong());
+        verify(userDaoMock, times(1)).findById(anyLong());
+        verify(userDaoMock, times(1)).update(any());
+        verifyNoMoreInteractions(userTeamDaoMock, teamDaoMock, userDaoMock);
+    }
+
+    @Test(expected = UserAlreadyInTeamException.class)
+    public void testJoinTeam_UserAlreadyInTeam() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+        given(userTeamDaoMock.checkIfUserTeamExists(anyLong(), anyLong())).willReturn(true);
+        service.joinTeam(99L, 1L);
+    }
+
+    @Test
+    public void testGetTeamMembersByTeamId_CorrectId() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+        assertEquals("3rd members ID should match with expected", 4, teamMembersDto.get(2).getId());
+
+        verify(teamDaoMock, times(1)).findById(anyLong());
+        verify(teamDaoMock, times(1)).getTeamMembersByTeam(any());
+        verifyNoMoreInteractions(teamDaoMock);
+        verifyZeroInteractions(userDaoMock, userTeamDaoMock);
+    }
+
+    @Test(expected = EntityNotFoundException.class)
+    public void testGetTeamMembersByTeamId_WrongId() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        given(teamDaoMock.findById(1L)).willThrow(new EntityNotFoundException());
+        service.getTeamMembersByTeamId(1L);
+    }
+
+    @Test
+    public void testFindByTeamName_CorrectName() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+        assertEquals(99L, teamDto.getId());
+        assertEquals("TestTeam", teamDto.getName());
+
+        verify(teamDaoMock, times(1)).findByName(anyString());
+        verifyNoMoreInteractions(teamDaoMock);
+        verifyZeroInteractions(userDaoMock, userTeamDaoMock);
+    }
+
+    @Test(expected = EntityNotFoundException.class)
+    public void testFindByTeamName_WrongName() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        given(teamDaoMock.findByName("TestTeam53")).willThrow(new EntityNotFoundException());
+        service.findByTeamName("TestTeam53");
+    }
+
+    @Test
+    public void testGetAllTeams() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: missing Javadoc, change method name like suggested in previous comment

**Comment by stanislavgrujic** [comment]
Please don't add Javadoc on test method names, it usually does not help readability. Rather try to name the methods so that the name indicates what is being tests in the method, e.g., `shouldFindEntityWithCorrectId`

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+    private TeamDao teamDaoMock;
+    @Mock
+    private UserDao userDaoMock;
+    @Mock
+    private UserTeamDao userTeamDaoMock;
+
+    @InjectMocks
+    private TeamServiceImpl service;
+
+    private static Team team() {
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: move all private methods under the public (Test) methods. It is hard to follow this way

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+    public void shouldFindTeamById() {
+        //given
+        given(teamDaoMock.findById(99)).willReturn(team());
+
+        //when
+        TeamDto teamDto = service.find(99L);
+
+        //then
+        assertNotNull(teamDto);
+        assertEquals("TestTeam", teamDto.getName());
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: swap argument order, usually, first should be object we are check, and second argument should be expected value. Apply this to all tests (if that should be applied).

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+
+    private static List<Team> teamList() {
+        List<Team> teamList = new ArrayList<>();
+        teamList.add(team());
+        teamList.add(new Team(100L, "TestTeam2", new HashSet<>()));
+        teamList.add(new Team(101L, "TestTeam3", new HashSet<>()));
+        return teamList;
+    }
+
+    @Test
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: Missing test javadoc. Usually, if the method name is descriptive, you can use something like: "See method name" or similar. This applies to all test methods.

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java (hunk 17 above)
Comment (ksajvan) [reply service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: ignore this one...my bad

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+    public void shouldFailToFindTeamById() {
+        given(teamDaoMock.findById(anyLong())).willThrow(new EntityNotFoundException("Team", 150L));
+        service.find(150L);
+    }
+
+    @Test
+    public void shouldCreateTeam() {
+        //given
+        given(teamDaoMock.checkIfTeamExists("TestTeam")).willReturn(false);
+        given(teamDaoMock.save(any())).willReturn(team());
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: can we just check if team really exists instead of mocking it?

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+        //then
+        assertNotNull(checkDTO);
+        assertEquals(99L, checkDTO.getId());
+        assertEquals("TestTeam", checkDTO.getName());
+    }
+
+    @Test(expected = TeamNameNotUniqueException.class)
+    public void shouldFailToCreateTeam() {
+        given(teamDaoMock.checkIfTeamExists("TestTeam")).willReturn(true);
+        service.createTeam("TestTeam");
```
Comment (ksajvan) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: I don't like this test. You should create team in //Given, then, try to create team with same name and expect that it will throw exception

File: repository/src/main/java/eu/execom/labs/pomodorospringreact/repository/UserDao.java [hunk 21]
```diff
@@ -2,6 +2,7 @@
 
 import eu.execom.labs.pomodorospringreact.entity.Team;
 import eu.execom.labs.pomodorospringreact.entity.User;
+import eu.execom.labs.pomodorospringreact.exception.EntityNotFoundException;
```
Comment (idostanic) [inline repository/src/main/java/eu/execom/labs/pomodorospringreact/repository/UserDao.java]: I see no changes in this PR to this file other than this import, therefore it should be an unused import?

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,360 @@
...
+    @InjectMocks
+    private TeamServiceImpl service;
+
+    /**
+     * Should successfully find Team by ID
+     */
+    @Test
+    public void shouldFindTeamById() {
+        //given
+        given(teamDaoMock.findById(99)).willReturn(team());
```
Comment (idostanic) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: What does "99" represent?

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+     * Should throw FailedToLeaveTeam exception, if team leave is not persisted in DB
+     */
+    @Test(expected = FailedToLeaveTeamException.class)
+    public void shouldFailToLeaveTeam() {
+        //given
+        User user = user();
+        Team team = team();
+        UserTeam userTeam = userTeam(user, team);
+
+        User userInTeam = new User(2, "test@test.test", "Mocker Testerson", null, new HashSet<>(), new HashSet<>());
```
Comment (idostanic) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: It makes sense to make use of builder pattern when creating a user, this can be an improvement point for now. We will discuss it afterwards

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 24]
```diff
@@ -0,0 +1,360 @@
...
+        //given
+        given(userTeamDaoMock.checkIfUserTeamExists(anyLong(), anyLong())).willReturn(true);
+
+        //when - then (throw UaitException)
+        service.joinTeam(99L, 1L);
+    }
+
+
+    /**
+     * See the method name
```
Comment (idostanic) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: If method name is self-explanatory, then the comment is unnecessary.

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+
+    private static List<Team> teamList() {
+        List<Team> teamList = new ArrayList<>();
+        teamList.add(team());
+        teamList.add(new Team(100L, "TestTeam2", new HashSet<>()));
+        teamList.add(new Team(101L, "TestTeam3", new HashSet<>()));
+        return teamList;
+    }
+}
+
```
Comment (idostanic) [inline service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: is there any cleanup mechanism for these test cases?

File: service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java (hunk 22 above)
Comment (ivanCosic) [reply service/src/test/java/eu/execom/labs/pomodorospringreact/service/TeamServiceTest.java]: It is just a random number, for mocking purposes.
